src/Disks/ObjectStorages/DiskObjectStorage.cpp

```diff
@@ -732,7 +732,11 @@ std::unique_ptr<ReadBufferFromFileBase> DiskObjectStorage::readFile(
         ? std::max<size_t>(settings.remote_fs_buffer_size, DBMS_DEFAULT_BUFFER_SIZE)
         : settings.remote_fs_buffer_size;
 
-    size_t total_objects_size = file_size ? *file_size : getTotalSize(storage_objects);
+    /// For object storage file_size and read_hint is interchangeable, and
+    /// read_hint should be respected, since it is passed by various MergeTree
+    /// readers, and we should not read more then this size, since this may
+    /// lead to excessive memory usage.
+    size_t total_objects_size = file_size ? *file_size : (read_hint ? *read_hint : getTotalSize(storage_objects));
```

Comment:
Doesn't it reset the purpose of `filesystem_cache_prefer_bigger_buffer_size`? `file_size` is a hard limit on the buffer size, but `read_hint` means only the size we actually will read, which I get why you added, but it will always make `filesystem_cache_prefer_bigger_buffer_size` unused, and its purpose is to avoid cache fragmentation and certain cache inefficiencies.

Reply:
> since read_buffer_creator does not have any arguments like buffer_size and relies on original read_hint/file_size anyway

the only allocation of the buffer is done by `AsynchronousBoundedReadBuffer`, other buffers in Matryoshka do not allocate anything and the buffer is passed through till the end of the buffers chain from async buffer.

Reply:
@kssenii you are right about it, but today @alesapin pointed me at `filesystem_cache_allow_background_download` (https://github.com/ClickHouse/ClickHouse/blob/7612a9c6686649b3cfd167b15ee9a3cf96cf7657/src/IO/ReadSettings.h#L61), that should avoid the problem with small segments in cache